extensions/common/iam/identity-trust/identity-trust-sts-embedded/src/main/java/org/eclipse/edc/iam/identitytrust/sts/embedded/EmbeddedSecureTokenService.java

```diff
@@ -0,0 +1,82 @@
+/*
+ *  Copyright (c) 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.iam.identitytrust.sts.embedded;
+
+import org.eclipse.edc.identitytrust.SecureTokenService;
+import org.eclipse.edc.jwt.spi.JwtDecorator;
+import org.eclipse.edc.jwt.spi.TokenGenerationService;
+import org.eclipse.edc.spi.iam.TokenRepresentation;
+import org.eclipse.edc.spi.result.Result;
+import org.jetbrains.annotations.Nullable;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Implementation of a {@link SecureTokenService}, that is capable of creating a self-signed ID token ("SI token") completely in-process.
+ * To that end, it makes use of the <a href="https://connect2id.com/products/nimbus-jose-jwt">Nimbus JOSE/JWT library</a>.<br/>
+ * As a recommendation, the private key it uses should not be used for anything else.
+ */
+public class EmbeddedSecureTokenService implements SecureTokenService {
+
+    private final TokenGenerationService tokenGenerationService;
+    private final Clock clock;
+    private final long validity;
+
+
+    public EmbeddedSecureTokenService(TokenGenerationService tokenGenerationService, Clock clock, long validity) {
+        this.tokenGenerationService = tokenGenerationService;
+        this.clock = clock;
+        this.validity = validity;
+    }
+
+    @Override
+    public Result<TokenRepresentation> createToken(Map<String, Object> claims, @Nullable String bearerAccessScope) {
+        Collection<JwtDecorator> decorators = new ArrayList<>();
+        decorators.add(new SelfIssuedTokenDecorator(claims, clock, validity));
+
+        var result = withAccessToken(claims, bearerAccessScope, decorators::add);
+
+        if (result.failed()) {
+            return result.mapTo();
+        }
+        return tokenGenerationService.generate(decorators.toArray(JwtDecorator[]::new));
```

Comment:
can be done in a more compact way no?
```java
return withAccessToken(claims, bearerAccessScope, decorators::add)
   .compose(v -> tokenGenerationService.generate(decorators.toArray(JwtDecorator[]::new)));
```

Reply:
Probably, still need to rework the PR before it's ready for review :)